lib/irb/type_completion/completor.rb

```diff
@@ -0,0 +1,239 @@
+            case t
+            when Types::SingletonType
+              return Types.class_name_of(t.module_or_class)
+            when Types::InstanceType
+              return Types.class_name_of(t.klass)
+            end
+          end
+          nil
+        end
+
+        case @result
+        in [:call_or_const, type, _name, _self_call]
+          call_or_const_doc.call type
+        in [:const, type, _name, scope]
+          if type
+            call_or_const_doc.call type
+          else
+            value_doc.call scope[name]
+          end
+        in [:gvar, _name, scope]
+          value_doc.call scope["$#{name}"]
+        in [:ivar, _name, scope]
+          value_doc.call scope["@#{name}"]
+        in [:cvar, _name, scope]
+          value_doc.call scope["@@#{name}"]
+        in [:call, type, _name, _self_call]
+          method_doc.call type
+        in [:lvar_or_method, _name, scope]
+          if scope.local_variables.include?(name)
+            value_doc.call scope[name]
+          else
+            method_doc.call scope.self_type
+          end
+        else
+        end
+      end
+
+      def candidates_from_result(result)
+        candidates = case result
+        in [:require, name]
+          retrieve_files_to_require_from_load_path
+        in [:require_relative, name]
+          retrieve_files_to_require_relative_from_current_dir
+        in [:call_or_const, type, name, self_call]
+          ((self_call ? type.all_methods : type.methods).map(&:to_s) - HIDDEN_METHODS) | type.constants
+        in [:const, type, name, scope]
+          if type
+            scope_constants = type.types.flat_map do |t|
+              scope.table_module_constants(t.module_or_class) if t.is_a?(Types::SingletonType)
+            end
+            (scope_constants.compact | type.constants.map(&:to_s)).sort
+          else
+            scope.constants.sort | ReservedWords
+          end
+        in [:ivar, name, scope]
+          ivars = scope.instance_variables.sort
+          name == '@' ? ivars + scope.class_variables.sort : ivars
+        in [:cvar, name, scope]
+          scope.class_variables
+        in [:gvar, name, scope]
+          scope.global_variables
+        in [:symbol, name]
+          Symbol.all_symbols.map { _1.inspect[1..] }
+        in [:call, type, name, self_call]
+          (self_call ? type.all_methods : type.methods).map(&:to_s) - HIDDEN_METHODS
+        in [:lvar_or_method, name, scope]
+          scope.self_type.all_methods.map(&:to_s) | scope.local_variables | ReservedWords
+        else
+          []
+        end
+        [name || '', candidates]
+      end
+
+      def analyze(code, binding = Object::TOPLEVEL_BINDING)
+        # Workaround for https://github.com/ruby/prism/issues/1592
+        return if code.match?(/%[qQ]\z/)
+
+        lvars_code = binding.local_variables.map do |name|
+          "#{name}="
+        end.join + "nil;\n"
+        code = lvars_code + code
+        ast = Prism.parse(code).value
```

Comment:
In the latest version of prism, you can pass in the surrounding scopes to the parse which should help with this. It would look something like:

```ruby
ast = Prism.parse(code, scopes: [binding.local_variables])
```

Reply:
Thank you for adding this option :+1:
updated

Reply:
Since we’re using a new API, let’s also add a version check to `build_completor`?

Reply:
Since we’re using a new API, let’s also add a version check to `build_completor`?

Reply:
Added. also add warn message why using `:type` completor failed.